### PR TITLE
Feature mailbox

### PR DESCRIPTION
**Pull request by dkochmanski**

### CODE REVIEW

**Comment by dkochmanski** [comment]
I've mentioned in the past that I want to implement this feature. I've reworked the code a little and here's the proposal.

File: test/bordeaux-threads-test.lisp [hunk 1]
```diff
@@ -273,6 +273,139 @@ Distributed under the MIT license (see LICENSE file)
...
+        (results-mailbox (make-mailbox))
+        ;; all errors in a thread should be send here. For
+        ;; intance if messages are not monotonous from the
+        ;; consumer perspective it is an ordering issue.
+        (errors-mailbox (make-mailbox))
+        ;; flag to tell consumers that we are full. We always
+        ;; have a timeout in make-1-consumer (and
+        ;; make-n-consumer returns immedietely), so we may
+        ;; safely assume that join-thread will return.
+        (konsument-je-żeby-jeść-p nil)
```
Comment (phoe) [inline test/bordeaux-threads-test.lisp:353]: > `(konsument-je-żeby-jeść-p nil)`

Nie sądzę, żeby nazwa tej zmiennej była zrozumiała dla osób niepolskojęzycznych.

Comment (dkochmanski) [reply test/bordeaux-threads-test.lisp:353]: it doesn't matter, it is clearly the variable indicating that processing has ended.

Comment (phoe) [reply test/bordeaux-threads-test.lisp:353]: I'd still rename it to English, even if just to avoid causing trouble for people who don't have Polish ogonkis on their keynaps. Symbol names in international code should be ASCII enough to be typable in the keyboard without copy-pasting.

**Comment by dkochmanski** [comment]
last commits fix the timeout for read-message (timeout decays now on spurious wakeups) and makes the producer put 5000 messages instead of 5000000 what puts a strain on the emulated/generic version.

**Comment by dkochmanski** [comment]
I'm marking this branch as wip due to failures on travis (can't reproduce them locally)

**Comment by dkochmanski** [comment]
the test was borked, because join-thread was not called for each thread.

Now all tests pass except for allegro which failed for other reason:
```
The internal data structures in the running Lisp image have been

corrupted and execution cannot continue.  Check all foreign functions

and any Lisp code that was compiled with high speed and/or low safety,

as these are two common sources of this failure.  If you cannot find

anything incorrect in your code you should contact technical support

for Allegro Common Lisp, and we will try to help determine whether

this is a coding error or an internal bug.

A core file can be produced at this time.  Core files may or may not

be useful in diagnosing an error of this type and are often not useful.

See doc/introduction.htm#core-dumps-2 for more details.

The command "curl -L https://raw.githubusercontent.com/sionescu/cl-travis/master/install.sh | sh" failed and exited with 1 during .

Your build has been stopped.
```

**Comment by sionescu** [comment]
A few requests before reviewing the code:
 * remove the SBCL implementation. Since mailboxes are not a primitive, it's unnecessary to use the host's implementation and I want Bordeaux-threads behaviour to be identical across CL implementations
 * squash the other two commits and capitalize the commit message

File: src/default-implementations.lisp [hunk 2]
```diff
@@ -318,6 +318,111 @@ cannot be decremented in that time, returns NIL without decrementing the count."
...
+    (zerop (%mailbox-count mailbox))))
+
+(defdfun receive-message (mailbox &key timeout)
+  "Removes the oldest message from MAILBOX and returns it as the primary
+value, and a secondary value of T. If MAILBOX is empty waits until a message
+arrives.
+
+If TIMEOUT is provided, and no message arrives within the specified interval,
+returns primary and secondary value of NIL."
+  (with-lock-held ((%mailbox-lock mailbox))
```
Comment (ralt) [inline src/default-implementations.lisp:368]: You need to care for the timeout there, don't you?

File: src/default-implementations.lisp [hunk 3]
```diff
@@ -318,6 +318,111 @@ cannot be decremented in that time, returns NIL without decrementing the count."
...
+  "Returns a fresh list containing all the messages in the
+mailbox. Does not remove messages from the mailbox."
+  (with-lock-held ((%mailbox-lock mailbox))
+    (loop
+      for (i . j) on (%mailbox-messages mailbox)
+      until (null j) collect i)))
+
+(defdfun mailbox-count (mailbox)
+  "Returns the number of messages currently in the mailbox."
+  (with-lock-held ((%mailbox-lock mailbox))
```
Comment (ralt) [inline src/default-implementations.lisp:353]: Do you need a lock to read a value?

**Comment by dkochmanski** [comment]
@sionescu removed sbcl, squashed and reworded

File: src/default-implementations.lisp (hunk 3 above)
Comment (dkochmanski) [reply src/default-implementations.lisp:353]: good question, I don't know..

File: src/default-implementations.lisp (hunk 2 above)
Comment (dkochmanski) [reply src/default-implementations.lisp:368]: good catch! looking into it.

Comment (dkochmanski) [reply src/default-implementations.lisp]: I've fixed both with a rather ugly solution. Semaphore fixup has a new commit while mailbox fixup is squashed to the main commit.

File: src/default-implementations.lisp [hunk 4]
```diff
@@ -293,31 +293,161 @@ T on success.
...
+    (%mailbox-count mailbox)))
+
+(defdfun mailbox-empty-p (mailbox)
+  "Returns true if MAILBOX is currently empty, NIL otherwise."
+  (with-lock-held ((%mailbox-lock mailbox))
```
Comment (ralt) [inline src/default-implementations.lisp:371]: Given that you take the lock here, there's the potential for needing a timeout as well, no?

File: src/default-implementations.lisp (hunk 2 above)
Comment (dkochmanski) [reply src/default-implementations.lisp:368]: I'll put it here for the reference

```
(defdfun wait-on-semaphore (semaphore &key timeout)
  "Decrement the count of SEMAPHORE by 1 if the count would not be negative.

Else blocks until the semaphore can be decremented. Returns generalized boolean
T on success.

If TIMEOUT is given, it is the maximum number of seconds to wait. If the count
cannot be decremented in that time, returns NIL without decrementing the count."
  (let ((deadline (and timeout
                       (+ (get-internal-real-time)
                          (* timeout internal-time-units-per-second))))
        (lock (%semaphore-lock semaphore)))
    (macrolet ((with-thunk (() &body body)
                 `(if timeout
                      (handler-case (with-timeout (timeout)
                                      (acquire-lock lock))
                        (timeout ()
                          nil)
                        (:no-error ()
                          (unwind-protect (progn ,@body)
                            (release-lock lock))))
                      (with-lock-held (lock)
                        ,@body))))
      (with-thunk ()
        (if (>= (%semaphore-counter semaphore) 1)
            (decf (%semaphore-counter semaphore))
            (progn
              ;; we need this loop because of a spurious wakeup possibility
              (loop until (>= (%semaphore-counter semaphore) 1)
                    do (cond
                         ((null (condition-wait (%semaphore-condition-variable semaphore)
                                                (%semaphore-lock semaphore)
                                                :timeout timeout))
                          (return-from wait-on-semaphore))
                         ;; unfortunately cv-wait may return T on timeout too
                         ((and deadline (>= (get-internal-real-time) deadline))
                          (return-from wait-on-semaphore))
                         (timeout
                          (setf timeout (/ (- deadline (get-internal-real-time))
                                           internal-time-units-per-second)))))
              (decf (%semaphore-counter semaphore))))))))
```

After a thought I don't think we should put a timeout above the lock. We do not block inside locks so they should finish very quickly (condition-wait releases the lock). This time may be neglected without a consequence I think.

@sionescu  what do you think?

**Comment by mdbergmann** [comment]
Interesting. I use a 'mailbox' in my library.

Please forgive the intrusion.
But I'm wondering. Is Bordeaux-Threads the right place to add such a feature?
A 'mailbox' is a relatively high-level application, so I'd say it should at least be in a separate package that is layer wise on top of the thread API.

Since a mailbox is an application, there maybe be different needs, like bounded, unbounded, synchronous, asynchronous, etc.

In [cl-gserver](https://github.com/mdbergmann/cl-gserver) the user can choose between the two. An unbounded mailbox can be quite memory intensive if you have millions of items in it.
I used 'cl-speedy-queue' with some custom locking for the bounded and lparallels cons-queue for the unbounded queue.
The lparallels unbounded queue is quite fast. I have some benchmarks in the readme of this project.

File: test/bordeaux-threads-test.lisp [hunk 5]
```diff
@@ -273,6 +273,138 @@ Distributed under the MIT license (see LICENSE file)
   (is (bt:semaphore-p (bt:make-semaphore)))
   (is (null (bt:semaphore-p (bt:make-lock)))))
 
+(test mailbox-send-receive
```
Comment (mdbergmann) [inline test/bordeaux-threads-test.lisp:276]: Shouldn't this be differentiated in different `test` blocks?

Comment (dkochmanski) [reply test/bordeaux-threads-test.lisp:276]: send and receive are paired, we'd have two similar blocks with `is` put in different places then

**Comment by dkochmanski** [comment]
I've modeled it after SBCL's mailbox in sb-concurrency, so it is meant as a generic off-the-shelf mailbox which is sometimes needed in applications with threads. Certainly it is not "one size fits them all" solution.

**Comment by mdbergmann** [comment]
> I've modeled it after SBCL's mailbox in sb-concurrency, so it is meant as a generic off-the-shelf mailbox which is sometimes needed in applications with threads. Certainly it is not "one size fits them all" solution.

Well. Is it adaptable/extendable so that it is generic?
If not then it is quite specific.

File: test/bordeaux-threads-test.lisp (hunk 5 above)
Comment (mdbergmann) [reply test/bordeaux-threads-test.lisp:276]: I know what you mean.
But to be honest, if there would be a problem, a test failure, say in line 321, I'd have to look at a lot of code to see what happened before.
This is hard to follow and maintain I suppose.

**Comment by dkochmanski** [comment]
Following the dictionary, I mean generic as `having no particularly distinctive quality or application`.

**Comment by mdbergmann** [comment]
Yeah. Sorry. As I said. I don't want to intrude.
I don't know the mailbox of sb-concurrency. Neither do I know why it is in SBCL and not added for general availability based on BT in the first place. I'm just saying there are already libraries out there which are based on BT and implement such a thing in various shapes.
It is good that there is something 'official' as part of BT. If it is flexible so people can adapt to their needs it will be used. I'm certain.

Are there benchmarks? I'd be interested whether I can use it as well.

**Comment by dkochmanski** [comment]
I remember comparing it against sbcl's /lockless/ implementation and afair it was around 1.5x slower. Other than that there are no benchmarks.

As of whether this should be part of bt or not, I'll leave the decision to @sionescu, I wrote it without any other dependencies than what bt provides and thought that it may be worth including in bt, but if it is not, I'm fine with that decision as well.

**Comment by sionescu** [comment]
@mdbergmann made some very good points. I was not aware of the two queue implementations that he mentioned and so I think that for the moment it doesn't make sense to add a mailbox to Bordeaux-threads.

It would be a good idea to have a library that implements various primitives for parallel programming and I see a few open questions to that end:

1. Can we make make such a library extensible, so that it provides some default implementations for bounded/unbounded queues, or thread pools, etc... but still allowing custom implementations ?
We could take inspiration from the JVM, Erlang, etc...
2. Do the CL implementations expose atomic operations that are sufficiently compatible to allow portably writing lockless queues ? I know there's [Shinmera/atomics](https://github.com/Shinmera/atomics) but I don't know how portable those primitives are in practice.

In any case, thanks for working on this.

**Comment by mdbergmann** [comment]
As a side note.
If one wanted the 'message-box' implementation of cl-gserver could be used directly which supports unbounded and bounded backing queues.
It's more or less just a matter of making an instance:
`(make-instance 'message-box-bt :max-queue-size max-queue-size)`

> 1. Can we make make such a library extensible, so that it provides some default implementations for bounded/unbounded queues, or thread pools, etc... but still allowing custom implementations ?
We could take inspiration from the JVM, Erlang, etc...

I find that `lparallel` is what this is, to a high degree.
For a one worker kernel there is too much overhead though, so I decided (after benchmarking) to make the message-box based on a single bare BT thread. This is ~10% faster than a one worker lparallel kernel.
However, for dispatching concurrent work lparallel is quite nice. The future has a design flaw (IMO) which was discussed in the latest lparallel ticket.

> 2. Do the CL implementations expose atomic operations that are sufficiently compatible to allow portably writing lockless queues ? I know there's Shinmera/atomics but I don't know how portable those primitives are in practice.

Are atomics really faster in a 'mailbox' scenario instead of locking? Remains to be seen.
transactional memory would also be an option. There is an excellent library out there: [stmx](https://stmx.org).
